#10 Fixing character logout save

UTengine opened the pull request:
2023-7-6 13:56, * Packet Count : recv=0, send=0, realsend=0 , time = 13:56 2023-7-6 13:57, SelectCharacter : acname=steve, name=REDACTED, index=0, pid : 59028, front : 8 2023-7-6 13:58, * Packet Count : recv=1, send=1, realsend=1 , time = 13:58 2023-7-6 14:0, * Packet Count : recv=1, send=1, realsend=1 , time = 14:0 2023-7-6 14:1, [Error-DB Fail] {call UPDATE_USER_DATA ( 'REDACTED', 1,1,101,0,0,0,15,789,0,0,0,0,0,132,132,100,75,65,60,50,50,1,42,615549,1,-13108,39960,155580,12680,0,?,?,?)}, Skill[] Item[] 2023-7-6 14:1, [AccountLogout] acname=steve
2023-7-6 14:1, Logout : steve, REDACTED (W:1,U:0)
2023-7-6 14:1, Invalid Logout : steve, REDACTED (W:1,U:0,A:1) 2023-7-6 14:13, UserDataLogout Check Error: steve, REDACTED (W:0) fixed state:
2023-7-6 14:40, * Packet Count : recv=7, send=7, realsend=7 , time = 14:40 2023-7-6 14:42, SelectCharacter : acname=steve, name=REDACTED, index=7, pid : 59028, front : 53 2023-7-6 14:42, [AccountLogout] acname=steve
2023-7-6 14:42, Logout : steve, REDACTED (W:1,U:1) 2023-7-6 14:42, SelectCharacter : acname=steve, name=REDACTED, index=8, pid : 59028, front : 60 2023-7-6 14:42, * Packet Count : recv=9, send=9, realsend=9 , time = 14:42

UTengine committed:
Fixing character logout save

2023-7-6 13:56, * Packet Count : recv=0, send=0, realsend=0 , time = 13:56
2023-7-6 13:57, SelectCharacter : acname=steve, name=REDACTED, index=0, pid : 59028, front : 8
2023-7-6 13:58, * Packet Count : recv=1, send=1, realsend=1 , time = 13:58
2023-7-6 14:0, * Packet Count : recv=1, send=1, realsend=1 , time = 14:0
2023-7-6 14:1, [Error-DB Fail] {call UPDATE_USER_DATA ( 'REDACTED', 1,1,101,0,0,0,15,789,0,0,0,0,0,132,132,100,75,65,60,50,50,1,42,615549,1,-13108,39960,155580,12680,0,?,?,?)}, Skill[] Item[]
2023-7-6 14:1, [AccountLogout] acname=steve
2023-7-6 14:1, Logout : steve, REDACTED (W:1,U:0)
2023-7-6 14:1, Invalid Logout : steve, REDACTED (W:1,U:0,A:1)
2023-7-6 14:13, UserDataLogout Check Error: steve, REDACTED (W:0)
fixed state:
2023-7-6 14:40, * Packet Count : recv=7, send=7, realsend=7 , time = 14:40
2023-7-6 14:42, SelectCharacter : acname=steve, name=REDACTED, index=7, pid : 59028, front : 53
2023-7-6 14:42, [AccountLogout] acname=steve
2023-7-6 14:42, Logout : steve, REDACTED (W:1,U:1)
2023-7-6 14:42, SelectCharacter : acname=steve, name=REDACTED, index=8, pid : 59028, front : 60
2023-7-6 14:42, * Packet Count : recv=9, send=9, realsend=9 , time = 14:42

stevewgr reviewed (changes requested):
Hi @UTengine, thanks for the PR.

I would like to avoid from merging this, although it "fixes" the issue. We have a ticket in the backlog to convert these columns to binary instead of varchar. Database had it right, but server files didn't. Please check: https://github.com/ko4life-net/ko/issues/94

srmeier commented on src/migration/10_fix_userdata_columns.sql:
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE [dbo].[USERDATA] ALTER COLUMN [strSkill] VARCHAR (20)
```
By default right now this is set to size 10, any reason why we should update this to 20?

stevewgr replied:
@UTengine, I agree with @srmeier here. The actual DBAgent C++ code reads it as 10 bytes.
Don't worry, I'll update this with additional changes after rebasing this PR.

stevewgr replied:
Done here: https://github.com/ko4life-net/ko-db/pull/12/commits/d5563e3b4cd371986eaa836dff6022ac03d016e7

UTengine replied:
There's a reason for it as it also did in official files, maybe someoone else knows I don't.

twostars replied:
> There's a reason for it as it also did in official files, maybe someoone else knows I don't.

10 is fine:
![image](https://github.com/ko4life-net/ko-db/assets/626562/72a9489b-1e66-4e8a-869f-fa2ce4da1a09)
![image](https://github.com/ko4life-net/ko-db/assets/626562/94573ddd-6a37-4b56-92c2-8188d91c19c5)

srmeier committed:
Rework migration scripts and convert columns to binary.

@srmeier:
+ Adjusting migration script numbers and updating the src/migration/10_fix_userdata_columns.sql script to update to BINARY
+ Adjusting column names to reflect datatype
+ updating the load user data proc

srmeier commented:
Wondering if we can marge this, along with [this](https://github.com/ko4life-net/ko/pull/214), as an incremental step?

stevewgr commented on import.ps1:
```diff
@@ -29,7 +29,7 @@ if ($server.Databases[$db_name]) {
 }
 
 foreach ($dir In $("schema", "procedure", "data", "migration")) {
-  $scripts = Get-ChildItem -Path src/$dir -Filter "*.sql"
+  $scripts = Get-ChildItem -Path src/$dir -Filter "*.sql" | Sort-Object { [regex]::Replace($_.Name, '\d+', { $args[0].Value.PadLeft(20) }) }
```
Good catch! Whilst this solution works, it may fail in other scenarios and hide potential errors. e.g. the fixed numeric value 20 for padding or if the prefixes will contain leading zeros.
Therefore and since we know the prefix have to start with a number and an underscore by convention, we can simplify this and be more explicit by using `_` as the split delimitir and cast it to an integer before sorting. That also enforces our intention to follow the convention, so that if the explicit cast fails, the user will see a clear error indicating there is something wrong with one of the migration scripts' names.
I'll add a fix for this in a separate PR where I refactor the import and export scripts. Something will this will do:
```powershell
function GetMigrationScripts {
  return Get-ChildItem "src/migration" -Filter "*.sql" | Sort-Object {[int]($_ -split '_')[0]} | ForEach-Object {$_}
}
```

stevewgr replied:
No longer relevant after rebasing the pr on master.

stevewgr commented on src/migration/9_fix_userdata_columns.sql:
```diff
@@ -0,0 +1,13 @@
+ALTER TABLE [dbo].[USERDATA] DROP CONSTRAINT [DF_USERDATA_strSkill]
+
+-- Implicit conversion from data type varchar to binary is not allowed.
+ALTER TABLE [dbo].[USERDATA] ADD [bSkill] BINARY (10) NULL;
+GO
+UPDATE [dbo].[USERDATA] SET [bSkill] = CONVERT(BINARY, [strSkill]);
+GO
+ALTER TABLE [dbo].[USERDATA] DROP COLUMN [strSkill];
+GO
```
Note that this will change the columns order in USERDATA table and put `bSkill` as the last column in the table. I know this is a cosmetic thing and we can always specify the order of the columns when we query it for data, but just for the sake of consistency, let me rework this and just recreate the table following the desired order.

stevewgr replied:
Done here: https://github.com/ko4life-net/ko-db/pull/12/commits/d5563e3b4cd371986eaa836dff6022ac03d016e7

stevewgr commented on src/migration/9_fix_userdata_columns.sql:
```diff
@@ -0,0 +1,13 @@
+GO
+ALTER TABLE [dbo].[USERDATA] DROP COLUMN [strSkill];
+GO
+EXEC sp_rename '[dbo].[USERDATA].[strItem]',  'bItem', 'COLUMN';
+GO
+EXEC sp_rename '[dbo].[USERDATA].[strSerial]',  'bSerial', 'COLUMN';
+GO
```
I'll also rework this, since it was my bad recommending on our Discord chat to rename the columns to `bVar` for bytes, whilst my intention was `byVar`.

stevewgr replied:
Also done here: https://github.com/ko4life-net/ko-db/pull/12/commits/d5563e3b4cd371986eaa836dff6022ac03d016e7

stevewgr reviewed (changes requested):
Awesome work @srmeier! Thanks for that.
I tested and your fix indeed works. I'll have to rebase your PR a bit based on my feedback below. Nothing that needs to be done from your part, this is just as FYI for myself and you.

stevewgr committed:
Run format on current work.

stevewgr committed:
Rename binary bVar columns to byVar.

+ Reworked the fix_userdata migration script to preserve columns order.

stevewgr committed:
Alter delete user data procedure and schema.

Both tables should now be compatible and use proper binary columns. Plus
the procedure will properly transfer data to DELETE_USERDATA table as a
backup resort.

stevewgr committed:
Convert warehouse columns to binary and rename them.

The rename is just for consistency with USERDATA table.
Also increase the capacity for bySerial column to match with byItem.

stevewgr committed:
Fix import script typo and skip formatting to speedup.

Note that when generating diffs, the code should already be formatted at
this point in the db and source, hence there is no need to format again,
which slows down things noticeably.

stevewgr committed:
Generate diffs for the migration scripts.

This will make it easier to see what they changed in the current db.
If you update any of these migration scripts, please generate the diffs
again so that they're synced.

stevewgr reviewed (approved):
Thanks guys for the patience and work! Time for shipping! 🚢 
![image](https://github.com/ko4life-net/ko-db/assets/90905675/07dd0b1c-536f-4eb6-85d0-64fffc1184b6)
